padl/transforms.py

```diff
@@ -1743,10 +1861,56 @@ def __call__(self, args):
         :param args: Arguments to call with.
         :return: Output from series of transforms.
         """
-        for transform_ in self.transforms:
-            args = transform_.pd_call_transform(args)
+        _in_args = args
+        for i, transform_ in enumerate(self.transforms):
+            try:
+                args = transform_.pd_call_transform(args)
+            except Exception as err:
+                self._pd_trace_error(i, _in_args)
+                raise err
         return args
 
+    def _pd_get_non_target_stage_idx(self, stage_used):
+        return 1 if self._pd_group else len(self)
+
+    def _pd_get_target_stage_idx(self, is_entire_transform: bool):
+        return 0 if self._pd_group and not is_entire_transform else _pd_trace[-1].error_position
+
+    def _pd_get_error_idx(self, stage: str):
+        """Track the index where a :class:`Compose` fails from the one that fails
+        on :meth:`self.pd_forward`.
+
+        Examples:
+            t = t_1 >> t_2 >> batch >> t_3 >> t_4
+            then,
+            t.pd_forward = t_3 >> t_4.
+            If we know that :meth:`t.pd_forward` is failing on `t_4`, which is its element 1, then
+            `t` is failing on len(t.pd_preprocess) + 1.
+
+            t = t_1 >> t_2 >> batch >> ((t_3 >> t_4) + (t_5 >> t_6))
+            then,
+            t.pd_forward = (t_3 >> t_4) + (t_5 >> t_6).
+            No matter what branch is failing on :meth:`t.pd_forward`, the error on `t` is on
+            len(t.pd_preprocess) + 0.
+        """
+        assert stage in ('forward', 'postprocess')
+        preprocess = self.pd_preprocess
+        forward = self.pd_forward
+        postprocess = self.pd_postprocess
+        preprocess_used = False if (isinstance(preprocess, Identity)) and \
+                                   not isinstance(self.__getitem__(0), Identity) else True
```

Comment:
Why can't we use the same check for `preprocess_used` as in `_pd_itercall`?

Reply:
It is not enough with `use_preprocess`, `use_forward` on the debugger is that in order to track properly where the error occurs when looking at the entire transform (not at the different stages). I explain it with a strange but allowed example. Let `t` be
```
t = prep >> batch >> unbatch >> post_1
```
and `t2` be:
```
t2 = prep >> batch >> Identity() >> unbatch >> post_1
```
and suppose an exception is got in `post_1`. When we are trying to track when the error happens considering the entire transform, we compute this:
```
len(t.pd_preprocess) + AtomicTransform()._pd_get_non_target_stage_idx() + index_failing_on_postprocess
```
Equation 1.
For both of them, `use_forward` is `True` because the statement for `use_forward` is:
```
use_forward = not isinstance(forward, Identity)
```
but one of them has to add a 1 on the second summand of  Equation 1 and the second has to add a 0

Reply:
Do you mean that `use_forward` is `False` for both?

Reply:
I think I see how this is different. But I don't think this check will always work. For example 
```python
import padl
from padl import transform, batch, unbatch, Identity
f = transform(lambda x: x + 1)
t = padl.Identity() >> f
preprocess_used = False if (isinstance(preprocess, Identity)) and not isinstance(t[0], Identity) else True
# prints True even though there is no preprocessing step
```

Reply:
> Do you mean that `use_forward` is `False` for both?

yes

Reply:
> I think I see how this is different. But I don't think this check will always work. For example
> 
> ```python
> import padl
> from padl import transform, batch, unbatch, Identity
> f = transform(lambda x: x + 1)
> t = padl.Identity() >> f
> preprocess_used = False if (isinstance(preprocess, Identity)) and not isinstance(t[0], Identity) else True
> # prints True even though there is no preprocessing step
> ```

right, so then maybe I should not consider this case and include this trivial `Identity`s on the ones considered on this issue: https://github.com/lf1-io/padl/issues/340

Reply:
This makes sense?

Reply:
It sounds like this could be solved by Issue #340 since resolving Issue #340 is awaiting the graph implementation of PADL.

Reply:
yes, so maybe we should remove this check, because it is also notworking on the 100% of cases

Reply:
removed and added as part of issue #340